config/crds/v1/bases/elasticsearch.k8s.elastic.co_elasticsearches.yaml

```diff
@@ -566,6 +560,8 @@ spec:
                               type: string
                           type: object
                         type: array
+                    required:
+                    - elasticsearchRefs
```

Comment:
This is unexpected

Reply:
Just so we are on the same page, the top level field `Monitoring` is set to optional:
https://github.com/elastic/cloud-on-k8s/blob/e9c6a259f46610fb5e14dc30c3e566b09d3ec8df/pkg/apis/elasticsearch/v1/elasticsearch_types.go#L135
The `Metrics` and `Logs` fields are also set to optional:
https://github.com/elastic/cloud-on-k8s/blob/e9c6a259f46610fb5e14dc30c3e566b09d3ec8df/pkg/apis/common/v1/monitoring.go#L11
And the `ElasticsearchRefs` for both of these are set to required:
https://github.com/elastic/cloud-on-k8s/blob/e9c6a259f46610fb5e14dc30c3e566b09d3ec8df/pkg/apis/common/v1/monitoring.go#L23

So maybe this is a bug fix that it's now doing what it's supposed to do? I'm still investigating, and looking through issues upstream, and will test to see if this causes any issues.

Reply:
Yep, here we go:
https://github.com/kubernetes-sigs/controller-tools/issues/943
https://github.com/kubernetes-sigs/controller-tools/pull/944
Here's what we have:
```
	// +kubebuilder:validation:Required
	ElasticsearchRefs []ObjectSelector `json:"elasticsearchRefs,omitempty"`
```
Which is both `omitempty` and `Required`, which used to be treated as optional, is now explicitly required.

Reply:
With something like the following, we get an expected failure:
```
spec:
  version: 8.15.0
  monitoring:
    metrics: {}
    logs: {}
```
Failure:
```
❯ kc apply -n default -f config/custom/elasticsearch-with-monitoring.yaml
elasticsearch.elasticsearch.k8s.elastic.co/elasticsearch-monitor-sink created
The Elasticsearch "elasticsearch-monitor-source" is invalid:
* spec.monitoring.logs.elasticsearchRefs: Required value
* spec.monitoring.metrics.elasticsearchRefs: Required value
```
And adjusting to this:
```
spec:
  version: 8.15.0
  monitoring:
    metrics:
      elasticsearchRefs:
      - name: elasticsearch-monitor-sink
        namespace: default
    logs:
      elasticsearchRefs:
      - name: elasticsearch-monitor-sink
        namespace: default
```
Gets the expected results:
```
❯ kc apply -n default -f config/custom/elasticsearch-with-monitoring.yaml
elasticsearch.elasticsearch.k8s.elastic.co/elasticsearch-monitor-sink unchanged
elasticsearch.elasticsearch.k8s.elastic.co/elasticsearch-monitor-source created
```

Reply:
Oh wait, from the failure in the smoke tests:
```
        	            	Elasticsearch.elasticsearch.k8s.elastic.co "es-apm-sample-n9bk" is invalid: [spec.monitoring.logs.elasticsearchRefs: Required value, spec.monitoring.metrics.elasticsearchRefs: Required value]
```
There must be another permutation that could cause issues... checking.

Reply:
So it must be the webhook doing this, as I'm running with just the built operator, without the webook, but see this in the logs when using the `es-apm-sample` we provide:
```
2024-08-21T08:57:46.026-0500    ERROR   manager.eck-operator    Reconciler error        {"service.version": "2.15.0-SNAPSHOT+def426fb", "controller": "elasticsearch-controller", "object": {"name":"es-apm-sample","namespace":"default"}, "namespace": "default", "name": "es-apm-sample", "reconcileID": "f1beddda-68ce-4b41-ad07-2459b3dede4e", "error": "[adjust resources: adjust discovery config: Elasticsearch.elasticsearch.k8s.elastic.co \"es-apm-sample\" is invalid: [spec.monitoring.logs.elasticsearchRefs: Required value, spec.monitoring.metrics.elasticsearchRefs: Required value], Elasticsearch.elasticsearch.k8s.elastic.co \"es-apm-sample\" is invalid: [spec.monitoring.logs.elasticsearchRefs: Required value, spec.monitoring.metrics.elasticsearchRefs: Required value]]", "errorCauses": [{"error": "adjust resources: adjust discovery config: Elasticsearch.elasticsearch.k8s.elastic.co \"es-apm-sample\" is invalid: [spec.monitoring.logs.elasticsearchRefs: Required value, spec.monitoring.metrics.elasticsearchRefs: Required value]"}, {"error": "Elasticsearch.elasticsearch.k8s.elastic.co \"es-apm-sample\" is invalid: [spec.monitoring.logs.elasticsearchRefs: Required value, spec.monitoring.metrics.elasticsearchRefs: Required value]"}]}
```

Reply:
@pebrc I've found 2 paths to resolve this, none of which I particularly like:
1. Set `monitoring.*.elasticsearchRefs` explicitly to optional, which was how it was being treated previously because of it also having `omitEmpty` in the struct tags.
2. Set `Elasticsearch.Spec.Monitoring` to be a pointer. Now this brings up another whole can of worms, as we're talking about changing APIs that are `V1`, which has major red flags *but* it works (along with a slew of other code changes checking for `nil`). Now note the [k8s api conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required) concerning optional fields, which they indicate should be a pointer, or have built-in nil values [maps/slices], so should these have been a pointer from the beginning?

I've tried other options, such as the following, to no avail:
* adding  `+optional` to the parent optional fields.
* removing `omitEmpty` from ElasticsearchRefs.

Reply:
I think making it `+optional` sounds about right to me, if this was effectively the previous behaviour. Making `Monitoring` a pointer would also be possible I think as long as the serialisation of the struct stays the same (which I think is the case) But a pointer to struct containing a slice feels kind of awkward, the nil/empty slice is the more idiomatic way of expresssing the absence of value. 

I do suggest however to run the full set of e2e tests on this PR before we merge this in case we missed another case.

Reply:
@kvalliyurnatt brought up a good point. In addition to setting this to optional, we could add some new validation such that the webhook could catch these being missing, and reject the change. I will also add this.

Reply:
Actually, this is more difficult than initially anticipated as we use an non-pointer for the monitoring struct, so there's no difference between the user defining this without any logs/metrics associations, and it just being empty, so this is a check we can't do @kvalliyurnatt .

Reply:
I vote for keeping the behaviour identical to the status quo (ignoring the difficulty you mentioned).